h5py/h5fd.pyx

```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
 # Flag for tracking allocation of space in file
 LOG_ALLOC     = H5FD_LOG_ALLOC      # 0x4000
 LOG_ALL       = H5FD_LOG_ALL        # (H5FD_LOG_ALLOC|H5FD_LOG_TIME_IO|H5FD_LOG_NUM_IO|H5FD_LOG_FLAVOR|H5FD_LOG_FILE_IO|H5FD_LOG_LOC_IO)
+
+
+# Implementation of 'fileobj' Virtual File Driver: HDF5 Virtual File
+# Layer wrapper over Python file-like object.
+# https://support.hdfgroup.org/HDF5/doc1.8/TechNotes/VFL.html
+
+# HDF5 events (read, write, flush, ...) are dispatched via
+# H5FD_class_t (struct of callback pointers, H5FD_fileobj_*). This is
+# registered as the handler for 'fileobj' driver via H5FDregister.
+
+# H5FD_fileobj_open callback acts first, expecting "filename" from
+# Python side in form
+#   hex(id(file)).encode('ASCII')
+# and retruning struct H5FD_fileobj_t (descendant of base H5FD_t)
+# which will hold file state. Other callbacks receive H5FD_fileobj_t
+# and operate on f.fileobj. If successful, callbacks must return zero;
+# otherwize non-zero value.
+
+
+# H5FD_t of file-like object
+ctypedef struct H5FD_fileobj_t:
+  H5FD_t base  # must be first
+  PyObject* fileobj
+  haddr_t eoa
+
+
+# A minimal subset of callbacks is implemented. Non-essential
+# parameters (dxpl, type) are ignored.
+
+from cpython cimport Py_INCREF, Py_DECREF
+from libc.stdint cimport *
+from libc.stdio cimport *
+
+
+cdef H5FD_fileobj_t *H5FD_fileobj_open(const char *name, unsigned flags, hid_t fapl, haddr_t maxaddr):
+    cdef unsigned long long int p
+    cdef H5FD_fileobj_t *f
+    if sscanf(name, "%llx", &p):
+        f = <H5FD_fileobj_t *>malloc(sizeof(H5FD_fileobj_t))
+        f.fileobj = <PyObject *>p
+        Py_INCREF(<object>f.fileobj)
+        f.eoa = 0
+        return f
+
+cdef herr_t H5FD_fileobj_close(H5FD_fileobj_t *f) except -1:
+    Py_DECREF(<object>f.fileobj)
+    free(f)
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eoa(const H5FD_fileobj_t *f, H5FD_mem_t type):
+    return f.eoa
+
+cdef herr_t H5FD_fileobj_set_eoa(H5FD_fileobj_t *f, H5FD_mem_t type, haddr_t addr) except -1:
+    f.eoa = addr
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eof(const H5FD_fileobj_t *f, H5FD_mem_t type) except -1:
+    (<object>f.fileobj).seek(0, SEEK_END)
+    return (<object>f.fileobj).tell()
+
+cdef herr_t H5FD_fileobj_read(H5FD_fileobj_t *f, H5FD_mem_t type, hid_t dxpl, haddr_t addr, size_t size, void *buf) except -1:
+    (<object>f.fileobj).seek(addr)
+    cdef b = (<object>f.fileobj).read(size)
+    if len(b) != size:
+        return 1
+    cdef unsigned char[:] mview = bytearray(b)
```

Comment:
another thought, some file-like objects support `readinto` which takes the allocated output and just fills it. This would be a big optimization for file-like objects that support this interface.

Reply:
My biggest problem with this code was to support older Cythons:
https://github.com/h5py/h5py/pull/1061#issuecomment-402138536
It should be measured, but I'm not really sure `readinto` would produce *major* speed-ups here. The actual file access is usually the bottleneck, not the `memcpy`.

Reply:
why do we need `bytearray` at all, shouldn't a `memoryview` work fine?

Reply:
This copy is happening _after_ any filters have been undone, which means it could be a large decompressed memcpy. `memoryview` (called `buffer` in py2) has been around since before 2.5, it's not a Cython feature. I was thinking just directly doing:
```C
python_view = PyMemoryView_FromMemory(buf, size, PyBUF_WRITE)
fileobj.readinto(python_view)
```

Reply:
But `PyMemoryView_FromMemory()` is only available in later Pythons, according to
https://docs.python.org/3/c-api/memoryview.html#c.PyMemoryView_FromMemory
It would be awesome if someone could help me with optimized (no-copy) and universal (across all supported Pythons, Cythons) variants of `H5FD_fileobj_read`, `H5FD_fileobj_write`.